resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodFactory.scala

```diff
@@ -209,9 +213,33 @@ private[spark] class ExecutorPodFactoryImpl(sparkConf: SparkConf)
         .build()
     }.getOrElse(executorContainer)
 
-    new PodBuilder(executorPod)
+    val (maybeSecretsMountedPod, maybeSecretsMountedContainer) =
+      mountSecretsBootstrap.map { bootstrap =>
+        bootstrap.mountSecrets(executorPod, containerWithLimitCores)
```

Comment:
So this is where my suggestion of getting rid of bootstraps and using inheritance breaks. For that to work, this code would have to use the same abstraction as the driver side (orchestrator + steps), but it doesn't.

Reply:
Yes, true.

Reply:
The submission client has to do more work in each step than the driver has to do to configure executors. The submission client's steps also set up accompanying Kubernetes resources and the driver's spark configuration, whereas the driver only has to use those Kubernetes resources to mount volumes into the pods.

Reply:
I'm not saying that they should be doing the same work, I'm saying that because they are not using the same abstraction, you can't use the same existing concepts here. You had to create the concept of a "bootstrap" to reuse code between the "driver orchestrator" and this code (the "executor orchestrator"), and kinda shoerhorn it into the whole thing so that you could share a little bit of code.

Whereas if both could reuse steps (like "initialize this init container"), you probably could avoid that altogether.

Reply:
The difference in the work involved is the reason why we need different abstractions between the two places. What's a cleaner abstraction that allows for this difference in workload?

Reply:
I've already made a ton of suggestions. I'm not asking you to modify this PR, I'm pointing out places where having a proper abstraction would clean up the code.

e.g. this is the place where calling a common "prepare this container" code taking a custom orchestrator for executors would help.

Otherwise, if tomorrow you need to add another step before starting driver and executors, you'll need another hack like "bootstraps" because they don't share the concept of an orchestrator that runs config steps.

Reply:
We can address the refactors in a follow-up - it would be good to get a refactor going pretty soon though since a lot of the future work will depend on the fundamental architecture here.

Reply:
I created https://issues.apache.org/jira/browse/SPARK-22839.